lib/fromZodIssue.ts

```diff
@@ -1,120 +1,36 @@
-  includePath: boolean;
-}): string {
-  const { issue, issueSeparator, unionSeparator, includePath } = props;
-
-  if (issue.code === 'invalid_union') {
-    return issue.unionErrors
-      .reduce<string[]>((acc, zodError) => {
-        const newIssues = zodError.issues
-          .map((issue) =>
-            getMessageFromZodIssue({
-              issue,
-              issueSeparator,
-              unionSeparator,
-              includePath,
-            })
-          )
-          .join(issueSeparator);
-
-        if (!acc.includes(newIssues)) {
-          acc.push(newIssues);
-        }
-
-        return acc;
-      }, [])
-      .join(unionSeparator);
-  }
-
-  if (issue.code === 'invalid_arguments') {
-    return [
-      issue.message,
-      ...issue.argumentsError.issues.map((issue) =>
-        getMessageFromZodIssue({
-          issue,
-          issueSeparator,
-          unionSeparator,
-          includePath,
-        })
-      ),
-    ].join(issueSeparator);
-  }
-
-  if (issue.code === 'invalid_return_type') {
-    return [
-      issue.message,
-      ...issue.returnTypeError.issues.map((issue) =>
-        getMessageFromZodIssue({
-          issue,
-          issueSeparator,
-          unionSeparator,
-          includePath,
-        })
-      ),
-    ].join(issueSeparator);
-  }
-
-  if (includePath && isNonEmptyArray(issue.path)) {
-    // handle array indices
-    if (issue.path.length === 1) {
-      const identifier = issue.path[0];
-
-      if (typeof identifier === 'number') {
-        return `${issue.message} at index ${identifier}`;
-      }
+export type FromZodIssueOptions =
+  | {
+      messageBuilder: MessageBuilder;
     }
-
-    return `${issue.message} at "${joinPath(issue.path)}"`;
-  }
-
-  return issue.message;
-}
-
-export type FromZodIssueOptions = {
-  issueSeparator?: string;
-  unionSeparator?: string;
-  prefix?: string | null;
-  prefixSeparator?: string;
-  includePath?: boolean;
-};
+  // maintain backwards compatibility
+  | Omit<CreateMessageBuilderProps, 'maxIssuesInMessage'>;
 
 export function fromZodIssue(
   issue: ZodIssue,
   options: FromZodIssueOptions = {}
 ): ValidationError {
-  const {
-    issueSeparator = ISSUE_SEPARATOR,
-    unionSeparator = UNION_SEPARATOR,
-    prefixSeparator = PREFIX_SEPARATOR,
-    prefix = PREFIX,
-    includePath = true,
-  } = options;
-
-  const reason = getMessageFromZodIssue({
-    issue,
-    issueSeparator,
-    unionSeparator,
-    includePath,
-  });
-  const message = prefixMessage(reason, prefix, prefixSeparator);
+  const messageBuilder = createMessageBuilderFromOptions(options);
+  const message = messageBuilder([issue]);
 
   return new ValidationError(message, { cause: new zod.ZodError([issue]) });
 }
+
+function createMessageBuilderFromOptions(
+  options: FromZodIssueOptions
+): MessageBuilder {
+  if ('messageBuilder' in options) {
+    return options.messageBuilder;
+  }
+
+  return createMessageBuilder(options);
+}
```

Comment:
[question] Would it make to DRY it out and move this inside the `MessageBuilder` file?

Reply:
Why would this be consider DRYer? I am missing the benefit.

Reply:
You have the exact same snippet in two places. Wouldn't it make the code DRYer if we move this to a central place? If one changes, the other one will have to change accordingly. Unless I'm mixing things up.

Reply:
Can you please suggest how this code could be changed to avoid repetition. How could it go into the `MessageBuilder`. I feel separation of concerns isn't right.